### PR TITLE
Shows how to support two build flavors: development and release.

### PR DESCRIPTION
**Pull request by pierredelisle**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -185,6 +185,35 @@ DEFINES += \
...
+# security and optimizations and 2) a "release" build which targets end-user devices
+# and disables debugging artifacts and logging, and enables security and optimizations.
+#
+# To facilitate supporting these two build flavors, the sample app source code supports
+# the Makefile variable BUILD_DEVELOPMENT. BUILD_DEVELOPMENT is enabled by default so that
+# with no command line argument pecified at compile-time, a development build is produced.
+# Specifying BUILD_DEVELOPMENT=0 when running make which will yield a "pseudo" release build.
+#
+#
+# IMPORTANT NOTE: we say "pseudo" release build because the sample app is never to be used
```
Comment (jaylogue) [inline Makefile]: Maybe reword slightly:

... is never to be used **as-is** as the firmware for an end-user device.

File: main/include/WeaveProjectConfig.h [hunk 2]
```diff
@@ -38,18 +40,35 @@
  */
 #define WEAVE_DEVICE_CONFIG_ENABLE_TEST_DEVICE_IDENTITY 34
 
-// Use a default pairing code if one hasn't been provisioned in flash.
+/**
```
Comment (jaylogue) [inline main/include/WeaveProjectConfig.h]: I don't think this should be a doxygen comment.  This instance of the #define isn't definitive w.r.t. documenting the value.  Therefore this should probably just be a simple comment.

File: main/include/WeaveProjectConfig.h [hunk 3]
```diff
@@ -38,18 +40,35 @@
...
+ */
 #define WEAVE_DEVICE_CONFIG_USE_TEST_PAIRING_CODE "NESTUS"
 
-// For convenience, enable Weave Security Test Mode and disable the requirement for
-// authentication in various protocols.
-//
-//    WARNING: These options make it possible to circumvent basic Weave security functionality,
-//    including message encryption. Because of this they MUST NEVER BE ENABLED IN PRODUCTION BUILDS.
-//
+/**
```
Comment (jaylogue) [inline main/include/WeaveProjectConfig.h]: Change to regular comment.

File: main/include/WeaveProjectConfig.h [hunk 4]
```diff
@@ -38,18 +40,35 @@
...
+ *
+ * These options make it possible to circumvent basic Weave security functionality,
+ * including message encryption.
+ * Because of this, the way these options are configured below
+ * MUST NEVER BE ENABLED IN PRODUCTION BUILDS.
+ */
 #define WEAVE_CONFIG_SECURITY_TEST_MODE 1
 #define WEAVE_CONFIG_REQUIRE_AUTH 0
 
+/**
```
Comment (jaylogue) [inline main/include/WeaveProjectConfig.h]: Change to regular comment.

File: main/include/WeaveProjectConfig.h [hunk 5]
```diff
@@ -144,6 +156,7 @@
  * importance level, this means all standard, critical, info and debug importance level
  * vi events get logged.
  */
+// FIXME --> define in Makefile?
```
Comment (jaylogue) [inline main/include/WeaveProjectConfig.h]: Can we remove this?

File: main/include/app_config.h [hunk 6]
```diff
@@ -16,6 +16,8 @@
  *    limitations under the License.
  */
 
+# FIXME: anything that should be different whether it is a development or release build?
```
Comment (jaylogue) [inline main/include/app_config.h]: Can we remove this?

File: Makefile (hunk 1 above)
Comment (pierredelisle) [reply Makefile]: done

File: main/include/WeaveProjectConfig.h (hunk 2 above)
Comment (pierredelisle) [reply main/include/WeaveProjectConfig.h]: fixed.

File: main/include/WeaveProjectConfig.h (hunk 3 above)
Comment (pierredelisle) [reply main/include/WeaveProjectConfig.h]: FIxed.

File: main/include/WeaveProjectConfig.h (hunk 4 above)
Comment (pierredelisle) [reply main/include/WeaveProjectConfig.h]: Fixed.

File: main/include/WeaveProjectConfig.h (hunk 5 above)
Comment (pierredelisle) [reply main/include/WeaveProjectConfig.h]: Fixed.

File: main/include/app_config.h (hunk 6 above)
Comment (pierredelisle) [reply main/include/app_config.h]: Fixed.

**Comment by pierredelisle** [comment]
Earlier comments addressed.

**Comment by jaylogue** [comment]
@pierredelisle Can you add similar constructs to the EFR32 and ESP32 apps as well?
